feat: Custom Messages Feature

Glubus opened the pull request:
---
name: Custom Messages Feature
about: Add custom messages system for water drinking reminders
title: "feat/custom-messages"
labels: enhancement, database, commands
---

# Feature Pull Request: Custom Messages System

## Feature Summary

Addition of a complete system allowing users to customize water drinking reminder messages. Users can now define their own messages for each frequency type (30 minutes, 1 hour, 3 hours) instead of using only default messages.

## Implementation Details

### Changes Made

#### Database
- **New `custom_messages` table** :
  - `id` : Auto-incremented unique identifier
  - `user_id` : ID of the user who added the message
  - `message_type` : Message type (`thirty_min`, `one_hour`, `three_hours`)
  - `message` : Custom message content
  - `created_at` : Creation date
  - `UNIQUE` constraint on `message_type` (one message per type)

#### New slash commands
- **`/add_msg`** : Add a custom message
  - Message type validation
  - Creator user registration
  - Automatic replacement if a message already exists
- **`/list_msg`** : List all custom messages
  - Display with creator user mention
  - Automatic truncation of long messages
- **`/remove_msg`** : Remove a custom message
  - Deletion by message type

#### System integration
- Modification of the reminder system to prioritize custom messages
- Automatic fallback to default messages if no custom message exists
- Global custom messages (not server-specific)

### New files created
- `apps/discord/src/commands/add_msg.rs`
- `apps/discord/src/commands/list_msg.rs`
- `apps/discord/src/commands/remove_msg.rs`

### Modified files
- `apps/discord/src/database.rs` : Added custom message management functions
- `apps/discord/src/reminder.rs` : Integration of custom messages in the reminder system
- `apps/discord/src/main.rs` : Registration of new commands
- `apps/discord/src/commands/mod.rs` : Added new modules

### Breaking Changes

- [x] No breaking changes
- [ ] Breaking changes (list below):
  - No breaking changes, existing system continues to work normally

## Testing

### Test Coverage

- [ ] Unit tests added/updated
- [ ] Integration tests added/updated
- [ ] E2E tests added/updated
- [x] Manual testing completed

### Manual tests performed
- [x] Adding custom messages with `/add_msg`
- [x] Listing messages with `/list_msg`
- [x] Removing messages with `/remove_msg`
- [x] Verifying fallback to default messages
- [x] Testing uniqueness constraint by message type

## Documentation

- [x] Code comments added
- [x] API documentation updated
- [x] User documentation updated
- [ ] README updated if needed
- [x] Changelog updated

### User documentation
New commands are documented with clear descriptions:
- `/add_msg` : "Add a custom message for water drinking reminders"
- `/list_msg` : "List custom messages for water drinking reminders"
- `/remove_msg` : "Remove a custom message for water drinking reminders"

## Checklist

- [x] Code follows project style guidelines
- [x] Feature flag implemented (if applicable) - Not applicable
- [x] Rollback plan considered - Removal of new commands and table if necessary

## Usage

### Usage examples
```bash
# Add a custom message for 30-minute reminders
/add_msg type:30 minutes message:"💧 Time to drink some water! 💧"

# List all custom messages
/list_msg

# Remove the custom message for 1-hour reminders
/remove_msg type:1 hour
```

### Behavior
1. **Priority** : Custom messages are used with priority
2. **Fallback** : If no custom message exists, default messages are used
3. **Global scope** : Custom messages apply to all users
4. **Uniqueness** : One custom message per frequency type

Commit:
feat: added custom message feature, can add remove or list

Commit:
feat: added custom message feature, can now add remove or list msg

Altaks commented on apps/discord/src/commands/list_msg.rs:
```diff
@@ -0,0 +1,92 @@
+                    .title("📝 Messages personnalisés");
+
+                for (msg_type, message, user_id) in messages {
+                    let display_type = match msg_type.as_str() {
+                        "thirty_min" => "30 minutes",
+                        "one_hour" => "1 heure",
+                        "three_hours" => "3 heures",
+                        _ => &msg_type,
+                    };
```
Move this to a small inline function

Altaks commented on apps/discord/src/reminder.rs:
```diff
@@ -1,25 +1,39 @@
+        ReminderFrequency::ThreeHours => "three_hours",
+    };
+
+    let content = if let Ok(Some(custom_msg)) = get_custom_message(message_type).await {
```
`unwrap_or` exists you know ?

Altaks reviewed (changes requested):
A lot of "Made with AI" gibberish i see... 
Fix it

Commit:
refactor: improve code style and fix lifetime issues

- Replace match statements with if let for better readability
- Simplify custom message fallback logic using match
- Improve error handling consistency across commands

Addresses PR feedback for cleaner, more idiomatic Rust code.

Glubus commented:
> Could you add a security layer such as :
> 
>     * Messages need to be validated by a moderator before being able to be used in global messages
> 
>     * The moderator should be defined by their used id
> 
>     * The moderator user id should be configurable in the `.env` file

Why not adding a specific role for some user ? such as ["user", "moderator", "admin"] by default an admin is set in .env and he can set maybe other moderator ?

Commit:
refactor : changed codes based on pull request review

Altaks commented:
[![thought](https://img.shields.io/badge/thought-6B7280)](https://pullpo.io/cc?l=thought) 
Inactivity
